### PR TITLE
feat: allow space group access

### PR DESCRIPTION
**Pull request by Shifan-Gu**
Closes: 
[8718](https://github.com/lightdash/lightdash/issues/8718)

### Description:

https://github.com/lightdash/lightdash/assets/101873365/90e675c0-82e2-4857-bf62-698cbf96dadd

### Reviewer actions

- [ ] I have manually tested the changes in the preview environment
- [ ] I have reviewed the code
- [ ] I understand that "request changes" will block this PR from merging

### CODE REVIEW

**Comment by rephus** [comment]
Hi @Shifan-Gu there are a few errors with types, could you take a look ?  I'll review the code and test it in the meantime 

image

File: packages/backend/src/controllers/spaceController.ts [hunk 1]
```diff
@@ -215,4 +216,73 @@ export class SpaceController extends BaseController {
             results: undefined,
         };
     }
+
+    /**
+     * Grant a user access to a space
```
Comment (rephus) [inline packages/backend/src/controllers/spaceController.ts:221]:
```suggestion
     * Grant a group access to a space
```

File: packages/backend/src/controllers/spaceController.ts [hunk 2]
```diff
@@ -215,4 +216,73 @@ export class SpaceController extends BaseController {
...
+     * @param projectUuid The uuid of the space's parent project
+     * @parmm spaceUuid The uuid of the space to update
+     * @param userUuid The uuid of the user to grant access to
```
Comment (rephus) [inline packages/backend/src/controllers/spaceController.ts:224]:
```suggestion
     * @param groupUuid The uuid of the group to grant access to
     * @param spaceRole The role for the group in the space
```

File: packages/backend/src/controllers/spaceController.ts [hunk 3]
```diff
@@ -215,4 +216,73 @@ export class SpaceController extends BaseController {
             results: undefined,
         };
     }
+
+    /**
+     * Grant a user access to a space
+     * @param projectUuid The uuid of the space's parent project
+     * @parmm spaceUuid The uuid of the space to update
```
Comment (rephus) [inline packages/backend/src/controllers/spaceController.ts:223]:
```suggestion
     * @param spaceUuid The uuid of the space to update
```

File: packages/backend/src/controllers/spaceController.ts [hunk 4]
```diff
@@ -215,4 +216,73 @@ export class SpaceController extends BaseController {
...
+                body.spaceRole,
+            );
+        return {
+            status: 'ok',
+            results: undefined,
+        };
+    }
+
+    /**
+     * Remove a user's access to a space
```
Comment (rephus) [inline packages/backend/src/controllers/spaceController.ts:258]:
```suggestion
     * Remove a group's access to a space
```

File: packages/backend/src/controllers/spaceController.ts [hunk 5]
```diff
@@ -215,4 +216,73 @@ export class SpaceController extends BaseController {
...
+     * @param projectUuid The uuid of the space's parent project
+     * @param spaceUuid The uuid of the space to update
+     * @param userUuid The uuid of the user to revoke access from
```
Comment (rephus) [inline packages/backend/src/controllers/spaceController.ts:261]:
```suggestion
     * @param groupUuid The uuid of the group to revoke access from
```

File: packages/backend/src/services/SpaceService/SpaceService.ts [hunk 6]
```diff
@@ -281,6 +281,66 @@ export class SpaceService extends BaseService {
         await this.spaceModel.removeSpaceAccess(spaceUuid, shareWithUserUuid);
     }
 
+    async addSpaceGroupAccess(
+        user: SessionUser,
+        spaceUuid: string,
+        shareWithGroupUuid: string,
+        spaceRole: SpaceMemberRole,
+    ): Promise<void> {
+        // TODO: check if user can access groups
```
Comment (rephus) [inline packages/backend/src/services/SpaceService/SpaceService.ts:290]: TODO ?

File: packages/backend/src/services/SpaceService/SpaceService.ts [hunk 7]
```diff
@@ -281,6 +281,66 @@ export class SpaceService extends BaseService {
...
+            spaceRole,
+        );
+    }
+
+    async removeSpaceGroupAccess(
+        user: SessionUser,
+        spaceUuid: string,
+        shareWithGroupUuid: string,
+    ): Promise<void> {
+        // TODO: check if user can access groups
```
Comment (rephus) [inline packages/backend/src/services/SpaceService/SpaceService.ts:320]: TODO ?

File: packages/common/src/utils/projectMemberRole.ts [hunk 8]
```diff
@@ -59,6 +81,24 @@ export const getHighestProjectRole = (
         undefined,
     );
 
+export const getHighestSpaceRole = (
+    spaceRoles: Array<SpaceMemberRole | undefined>,
+): SpaceMemberRole | undefined =>
+    spaceRoles.reduce<SpaceMemberRole | undefined>((highestRole, role) => {
```
Comment (rephus) [inline packages/common/src/utils/projectMemberRole.ts:87]: is it not easier to `filter(!undefined).sortBy(SpaceRoleOrder).reverse()[0]`

**Review by rephus — CHANGES_REQUESTED** [review]
The implementation looks good, most of the code seems a copy from projectGroups. There were a few typos on the copy-paste in the controller. 

I'd like to test this once the `type` errors are fixed.

**Comment by Shifan-Gu** [comment]
Thank you @rephus ! 
I will work on the commented issue.

**Review by rephus — COMMENTED** [review]
It woks pretty well, There is only one UI thing we should fix to avoid confusion, but it can be done on a separate ticket

If I add a user , there is a warning message that appears if the user does not have enough `project` permissions to perform actions in the space

Screenshot from 2024-05-21 08-41-16

But When I add a group, this message doesn't appear, and I can't make changes in the space (so I don't have edit permissions) 
Screenshot from 2024-05-21 08-41-35


Apart from that, everything seems to be working fine, and it is backwards compatible.

**Comment by Shifan-Gu** [comment]
Hi @rephus

I think what's happening for the second case is that you give the viewer group a space editor role, and it overrides the project viewer role so you don't see any warning here.

I will work on this ticket https://github.com/lightdash/lightdash/issues/9539 which would help explain the role and permission lineage.

**Comment by rephus** [comment]
Sorry, this is taking longer to review because of this change https://github.com/lightdash/lightdash/pull/10082/commits/0d72b169ecdcce95c89c4727afa06783bbfbd69e I need to test everything again. Next time I'd rather do these non-blocking changes on a separate PR

**Comment by rephus** [comment]
The issue I mentioned hasn't been fixed. I don't get any warnings if the permission is not correct.

Screenshot from 2024-05-23 09-43-39

But let's not fix this here, let's merge as it is and we'll fix it on a separate PR

**Comment by rephus** [comment]
Another weird UI behaviour, probably caused by the recent change in permissions. 
I can upgrade permissions for indivdiual users in the group, but I can downgrade. 

eg: if The group can `edit` I can't set an idividual user to `viewer`


[Screencast from 23-05-24 09:50:54.webm](https://github.com/lightdash/lightdash/assets/1983672/60979ae1-0e2a-406c-9682-4412c8ead633)

Again, let's not fix it here, I'll create a separate issue.

File: packages/backend/src/controllers/spaceController.ts (hunk 4 above)
Comment (rephus) [reply packages/backend/src/controllers/spaceController.ts:259]: still not done

**Comment by rephus** [comment]
Waiting for tests to run. I tested manually and it seems good

**Review by rephus — APPROVED** [review]

**Pull request merged**